pubpub/platform
# feat: site builder api

This PR adds a site builder API, alongside an access token based authentication system. 

## Issue(s) Resolved
Resolves #366 

## Test Plan

1. Log in as a superadmin
2. Go to /c/croccroc/settings/tokens

From here on you'll create a number of tokens, please write them down somewhere

3. Create
    - A token with only community read permissions, call it Nothing 
    - A token with all permissions, call it All
    - A token with all permissions, call it "Only submitted" but for 'Pub:write' do the following. Click on `write` for Pub, you'll see a Popup come up. Unselect all stages except "Submitted".
    - Optional: a stage read token, do the same as for the "Only submitted" token above, but for "Stage:read" instead. Call it "Only submitted read"

4. Using the `All` token in the format `Bearer ${token}` in the Authorization header, find out the following
   - The `id` of the "Submitted" stage by doing `GET /api/v0/c/croccroc/stages`. Also write down another stage's id (does not really matter which one)
   - The `id` of the "Submission" pub type by doing `GET /api/v0/c/croccroc/pub-types`

5. Using the `All` token, try to create a nested pub by doing `POST /api/v0/c/croccroc/pubs` with this body
```json
{
  "pubTypeId": "{Submission pub type id}",
  "stageId": "{Submitted stage id}",
  "values": {
    "pubpub:title": "I am parent",
    "pubpub:content": "I am full of knowledge"
  },
  "children": [
    {
      "pubTypeId": "{Submission pub type id}",
      "values": {
        "pubpub:title": "I am child 1"
      },
      "children": [
        {
          "pubTypeId": "{Submission pub type id}",
          "values": {
            "pubpub:title": "I am a grandchild"
          }
        }
      ]
    },
    {
      "pubTypeId": "{Submission pub type id}",
      "values": {
        "pubpub:title": "I am child 2"
      }
    }
  ]
}
```

This should work

6. using the All token, try `GET /api/v0/c/croccro/pubs` and observe that all the pubs and children are there

Ideally, try to repeat the above with the other tokens, and see the error messages.

Sorry I don't have any more time to improve this, let me know if this is insufficient information to test it!
And also sorry, these should be automated tests, but I don't know how far our testing infra is yet.

## Screenshots (if applicable)

The access token management screen

![image](https://github.com/pubpub/v7/assets/21983833/61cbdbbb-0b6a-4db3-8513-b6671871df7e)

Token creation flow with error states


https://github.com/pubpub/v7/assets/21983833/2beb473b-16a0-4f81-90bf-9df62f01d35d



## Optional

### Notes/Context/Gotchas

The idea behind the access token permissions is roughly this:

An access token has Permissions.

A permission is basically a tuple of
- scope (which routes it can touch, `pub`, `stage`, `community`, `member` etc)
- accessType (what kind of operation it can do, `read`, `write`, `archive` (unsure whether archive needs to be a separate thing)
- constraints. Some permissions can have extra constraints, such as only being able to access certain things. For the `pub` scope for instance, you can select which stages you can write a pub to.


#### MISSING

Things that are not yet done

- [x] Removing tokens
- [ ] Read members
- [ ] Read community information
- [ ] for `/stages`, only allow reading of stages you have access to


### Supporting Docs
